### PR TITLE
fix: DotClass prop changes in Carousel

### LINKED ISSUE #21131: Not able to forward dotClass prop in carousel
- [x] I have searched the [issues](https://github.com/ant-design/ant-design/issues) of this repository and believe that this is not a duplicate.

### Reproduction link
[Edit on CodeSandbox](https://codesandbox.io/s/polished-morning-lb2qm)

### Steps to reproduce
1) Pass custom dot class name into a carousel.
2) Not able to forward that class name
3) " const dotsClass = 'slick-dots';
    const dotPosition = this.getDotPosition();
    props.vertical = dotPosition === 'left' || dotPosition === 'right';
    props.dotsClass = `${dotsClass} ${dotsClass}-${dotPosition || 'bottom'}`;"

### What is expected?
1) " const dotsClass = 'slick-dots';
    const dotPosition = this.getDotPosition();
    props.vertical = dotPosition === 'left' || dotPosition === 'right';
    props.dotsClass = `${dotsClass} ${dotsClass}-${dotPosition || 'bottom'} ${props.dotsClass}`;"

### What is actually happening?
Not able to forward a custom dot class name

| Environment | Info |
|---|---|
| antd | 3.26.5 |
| React | 16.3.0 |
| System | Mac |
| Browser | Chrome |

### PR DESCRIPTION
**Pull request by krishangoelmt**
### 🤔 This is a ...

- [ ] New feature
- [x] Bug fix
- [ ] Site / document update
- [ ] Component style update
- [ ] TypeScript definition update
- [x] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 🔗 Related issue link

close #21131

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |           |
| 🇨🇳 Chinese |           |

### ☑️ Self Check before Merge

- [x] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

-----
[View rendered components/carousel/index.en-US.md](https://github.com/krishangoelmt/ant-design/blob/master/components/carousel/index.en-US.md)

### CODE REVIEW

File: components/carousel/index.tsx [hunk 1]
```diff
@@ -21,13 +21,15 @@ export interface CarouselProps extends Settings {
...
   children?: React.ReactNode;
+  dotsClass?: string;
 }
 
 export default class Carousel extends React.Component<CarouselProps, {}> {
   static defaultProps = {
     dots: true,
     arrows: false,
     draggable: false,
+    dotsClass: '',
```
Comment (yoyo837) [inline components/carousel/index.tsx]: default null?

Comment (krishangoelmt) [reply components/carousel/index.tsx]: this is the fallback condition

Comment (krishangoelmt) [reply components/carousel/index.tsx]: Let me know when can I expected to be merge.

Comment (krishangoelmt) [reply components/carousel/index.tsx]: Also, let me know if any changes required from my side in this PR.

Comment (afc163) [reply components/carousel/index.tsx]: image

Comment (krishangoelmt) [reply components/carousel/index.tsx]: all issue resolved

File: scripts/azure-github-comment.js [hunk 2]
```diff
@@ -33,10 +33,8 @@ async function withGithub(url, json, method) {
 
   // Find my comment
   const updateComment = comments.find(({ body }) => body.includes(REPLACE_MARK));
-  console.log('Origin comment:', updateComment);
 
   // Update
-  let res;
```
Comment (yoyo837) [inline scripts/azure-github-comment.js]: Why delete this?

Comment (krishangoelmt) [reply scripts/azure-github-comment.js]: Please check again.

Comment (yoyo837) [reply scripts/azure-github-comment.js]: `components/carousel/index.tsx`
  114:7  error  Unexpected string concatenation  prefer-template

Please use ES6 template string.

**Comment by yoyo837** [comment]
`components/carousel/index.tsx`
  114:7  error  Unexpected string concatenation  prefer-template

Please use ES6 template string.

File: components/carousel/index.tsx [hunk 3]
```diff
@@ -106,7 +108,11 @@ export default class Carousel extends React.Component<CarouselProps, {}> {
     const dotsClass = 'slick-dots';
     const dotPosition = this.getDotPosition();
     props.vertical = dotPosition === 'left' || dotPosition === 'right';
-    props.dotsClass = `${dotsClass} ${dotsClass}-${dotPosition || 'bottom'}`;
+    props.dotsClass = classNames(
+      dotsClass,
+      this.props.dotsClass,
+      `${dotsClass}-${dotPosition || 'bottom'}`,
+    );
```
Comment (yoyo837) [inline components/carousel/index.tsx]: @afc163 `dotsClass` Should it be replaced or supplemented？🤔

Comment (krishangoelmt) [reply components/carousel/index.tsx]: it's same as the previous one.

Comment (yoyo837) [reply components/carousel/index.tsx]: Your changes is to supplement it, not replace it.

Maybe it should be like this:
```js
props.dotsClass = classNames(
    this.props.dotsClass,
    `${this.props.dotsClass}-${dotPosition || 'bottom'}`,
);
```
And then provide all the styles yourself need. 

My personal opinion.

Comment (krishangoelmt) [reply components/carousel/index.tsx]: The reason behind doing this is we don't want to lose the whole style provided by slick slider. By simply applying an external class name we can overwrite that style easily.

Comment (yoyo837) [reply components/carousel/index.tsx]: Ok, Let the owner to decide it.

Comment (krishangoelmt) [reply components/carousel/index.tsx]: ok

**Comment by zombieJ** [comment]
We can reuse `dots` prop:
```tsx
type dots  = boolean | {
  className?: string;
};
```

File: components/carousel/index.tsx (hunk 3 above)
Comment (zombieJ) [reply components/carousel/index.tsx]: I prefer patch additional className instead of modify origin provided

File: components/carousel/index.tsx [hunk 4]
```diff
@@ -106,7 +108,11 @@ export default class Carousel extends React.Component<CarouselProps, {}> {
     const dotsClass = 'slick-dots';
     const dotPosition = this.getDotPosition();
     props.vertical = dotPosition === 'left' || dotPosition === 'right';
-    props.dotsClass = `${dotsClass} ${dotsClass}-${dotPosition || 'bottom'}`;
+    props.dotsClass = classNames(
+      dotsClass,
+      this.props.dotsClass,
```
Comment (afc163) [inline components/carousel/index.tsx]:
```suggestion
      props.dotsClass,
```

**Comment by yoyo837** [comment]
> We can reuse `dots` prop:
> 
> ```tsx
> type dots  = boolean | {
>   className?: string;
> };
> ```

Cannot be compatible with upstream ts types

**Comment by zombieJ** [comment]
> Cannot be compatible with upstream ts types

It's OK since we have list `dots` in our document. And we can override this prop.

**Comment by yoyo837** [comment]
> 
> It's OK since we have list `dots` in our document. And we can override this prop.

Without interface extends?

**Comment by zombieJ** [comment]
extends the dots

File: components/carousel/index.tsx (hunk 3 above)
Comment (krishangoelmt) [reply components/carousel/index.tsx]: Do I make any changes in this pull request or not?

Comment (krishangoelmt) [reply components/carousel/index.tsx]: any update about this PR?

Comment (yoyo837) [reply components/carousel/index.tsx]: https://github.com/ant-design/ant-design/pull/21156#issuecomment-581262408

File: components/carousel/index.tsx [hunk 5]
```diff
@@ -12,6 +12,7 @@ const SlickCarousel = require('@ant-design/react-slick').default;
 
 export type CarouselEffect = 'scrollx' | 'fade';
 export type DotPosition = 'top' | 'bottom' | 'left' | 'right';
+export type dots = boolean | { className?: string };
```
Comment (zombieJ) [inline components/carousel/index.tsx]: `dots` => `Dots`

File: components/carousel/index.en-US.md [hunk 6]
```diff
@@ -23,6 +23,7 @@ A carousel component. Scales with its container.
 | dots | Whether to show the dots at the bottom of the gallery | boolean | `true` |  |
 | easing | Transition interpolation function name | string | `linear` |  |
 | effect | Transition effect | `scrollx` \| `fade` | `scrollx` |  |
+| dotsClass | Add custom className for dots | string | - |  |
```
Comment (zombieJ) [inline components/carousel/index.en-US.md]: Doc should update

Comment (krishangoelmt) [reply components/carousel/index.en-US.md]: what should I need to update here plz guide me

Comment (zombieJ) [reply components/carousel/index.en-US.md]: Remove `dotClass` and update `dots` to  `boolean | { className: string }`

**Pull request closed without merging**